`Vec`'s const_make_global method erroring when `Vec` is empty

I tried this code:

```rust
#![feature(const_heap)]
fn main() {
    const {
        let mut inner_vec = Vec::<i32>::new();
        let inner_slice = inner_vec.const_make_global();
    }
}
```

I expected `const_make_global()` to just return an empty `'static` slice

Instead, it gave me this error:

```
error[E0080]: pointer not dereferenceable: pointer must point to some allocation, but got 0x4[noalloc] which is a dangling pointer (it has no provenance)
   --> src\main.rs:6:27
    |
  6 |         let inner_slice = inner_vec.const_make_global();
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `main::{constant#0}` failed inside this call
    |
note: inside `Vec::<i32>::const_make_global`
   --> C:\Users\theiz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\vec\mod.rs:897:18
    |
897 |         unsafe { core::intrinsics::const_make_global(self.as_mut_ptr().cast()) };
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the failure occurred here
```

But this code:

```rust
#![feature(const_heap)]
#![feature(const_trait_impl)]
fn main() {
    const {
        let mut inner_vec = Vec::<i32>::new();
        inner_vec.push(4);
        let inner_slice = inner_vec.const_make_global();
    }
}
```

works just fine


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.95.0-nightly (859951e3c 2026-02-24)
binary: rustc
commit-hash: 859951e3c7c9d0322c39bad49221937455bdffcd
commit-date: 2026-02-24
host: x86_64-pc-windows-msvc
release: 1.95.0-nightly
LLVM version: 22.1.0
```

I am assuming it was because the `Vec` was empty

--- a/library/alloc/src/vec/mod.rs
+++ b/library/alloc/src/vec/mod.rs
@@ -898,6 +898,20 @@ impl<T> Vec<T> {
     where
         T: Freeze,
     {
+        // const_make_global requires the pointer to point to the beginning of a heap allocation,
+        // which is not the case when `self.capacity()` is 0, which is why we instead return a new slice.
+        if self.capacity() == 0 {
+            _ = ManuallyDrop::new(self);
+            return &[];
+        }
+
+        // if `T::IS_ZST`, the pointer we give to `const_make_global` would again not
+        // point to the start of a heap allocation
+        if T::IS_ZST {
+            let me = ManuallyDrop::new(self);
+            return unsafe { slice::from_raw_parts(NonNull::<T>::dangling().as_ptr(), me.len) };
+        }
+
         unsafe { core::intrinsics::const_make_global(self.as_mut_ptr().cast()) };
         let me = ManuallyDrop::new(self);
         unsafe { slice::from_raw_parts(me.as_ptr(), me.len) }

--- a/library/alloc/src/vec/mod.rs
+++ b/library/alloc/src/vec/mod.rs
@@ -898,23 +898,17 @@ impl<T> Vec<T> {
     where
         T: Freeze,
     {
-        // const_make_global requires the pointer to point to the beginning of a heap allocation,
-        // which is not the case when `self.capacity()` is 0, which is why we instead return a new slice.
-        if self.capacity() == 0 {
-            _ = ManuallyDrop::new(self);
-            return &[];
-        }
-
-        // if `T::IS_ZST`, the pointer we give to `const_make_global` would again not
-        // point to the start of a heap allocation
-        if T::IS_ZST {
+        // `const_make_global` requires the pointer to point to the beginning of a heap allocation,
+        // which is not the case when `self.capacity()` is 0, or if `T::IS_ZST`,
+        // which is why we instead return a new slice in this case.
+        if self.capacity() == 0 || T::IS_ZST {
             let me = ManuallyDrop::new(self);
-            return unsafe { slice::from_raw_parts(NonNull::<T>::dangling().as_ptr(), me.len) };
+            unsafe { slice::from_raw_parts(NonNull::<T>::dangling().as_ptr(), me.len) }
+        } else {
+            unsafe { core::intrinsics::const_make_global(self.as_mut_ptr().cast()) };
+            let me = ManuallyDrop::new(self);
+            unsafe { slice::from_raw_parts(me.as_ptr(), me.len) }
         }
-
-        unsafe { core::intrinsics::const_make_global(self.as_mut_ptr().cast()) };
-        let me = ManuallyDrop::new(self);
-        unsafe { slice::from_raw_parts(me.as_ptr(), me.len) }
     }
 }
 

--- a/library/alloctests/tests/vec.rs
+++ b/library/alloctests/tests/vec.rs
@@ -2764,3 +2764,25 @@ fn const_heap() {
 
     assert_eq!([1, 2, 4, 8, 16, 32], X);
 }
+
+// regression test for issue #153158. `const_make_global` previously assumed `Vec<T>`'s buf
+// always has a heap allocation, which lead to compilation errors.
+#[test]
+fn const_make_global_empty_or_zst_regression() {
+    const EMPTY_SLICE: &'static [i32] = {
+        let empty_vec: Vec<i32> = Vec::new();
+        empty_vec.const_make_global()
+    };
+
+    assert_eq!(EMPTY_SLICE, &[]);
+
+    const ZST_SLICE: &'static [()] = {
+        let mut zst_vec: Vec<()> = Vec::new();
+        zst_vec.push(());
+        zst_vec.push(());
+        zst_vec.push(());
+        zst_vec.const_make_global()
+    };
+
+    assert_eq!(ZST_SLICE, &[(), (), ()]);
+}

